cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,197 @@
+
+	f.StringArrayVarP(&copyOptions.Hosts, "host", "H", nil, "only consider snapshots for this `host`, when no snapshot ID is given (can be specified multiple times)")
+	f.Var(&copyOptions.Tags, "tag", "only consider snapshots which include this `taglist`, when no snapshot-ID is given")
+	f.StringArrayVar(&copyOptions.Paths, "path", nil, "only consider snapshots which include this (absolute) `path`, when no snapshot-ID is given")
+}
+
+func runCopy(opts CopyOptions, gopts GlobalOptions, args []string) error {
+	if opts.Repo == "" {
+		return errors.Fatal("Please specify a destination repository location (-repo2)")
+	}
+	var err error
+	dstGopts := gopts
+	dstGopts.Repo = opts.Repo
+	dstGopts.PasswordFile = opts.PasswordFile
+	dstGopts.PasswordCommand = opts.PasswordCommand
+	dstGopts.KeyHint = opts.KeyHint
+	dstGopts.password, err = resolvePassword(dstGopts, "RESTIC_PASSWORD2")
+	if err != nil {
+		return err
+	}
+	dstGopts.password, err = ReadPassword(dstGopts, "enter password for destination repository: ")
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(gopts.ctx)
+	defer cancel()
+
+	srcRepo, err := OpenRepository(gopts)
+	if err != nil {
+		return err
+	}
+
+	dstRepo, err := OpenRepository(dstGopts)
+	if err != nil {
+		return err
+	}
+
+	srcLock, err := lockRepo(srcRepo)
+	defer unlockRepo(srcLock)
+	if err != nil {
+		return err
+	}
+
+	dstLock, err := lockRepo(dstRepo)
+	defer unlockRepo(dstLock)
+	if err != nil {
+		return err
+	}
+
+	debug.Log("Loading source index")
+	if err := srcRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	debug.Log("Loading destination index")
+	if err := dstRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Hosts, opts.Tags, opts.Paths, args) {
+		Verbosef("snapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+		Verbosef("  copy started, this may take a while...\n")
+
+		if err := copySnapshot(ctx, srcRepo, dstRepo, *sn.Tree); err != nil {
+			return err
+		}
+		debug.Log("tree copied")
+
+		if err = dstRepo.Flush(ctx); err != nil {
+			return err
+		}
+		debug.Log("flushed packs")
+
+		err = dstRepo.SaveIndex(ctx)
+		if err != nil {
+			debug.Log("error saving index: %v", err)
+			return err
+		}
+		debug.Log("saved index")
+
+		// save snapshot
+		sn.Parent = nil   // Parent does not have relevance in the new repo.
+		sn.Original = nil // Original does not have relevance in the new repo.
+		newID, err := dstRepo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+		if err != nil {
+			return err
+		}
+		Verbosef("snapshot %s saved\n", newID.Str())
+	}
+	return nil
+}
+
+func copySnapshot(ctx context.Context, srcRepo, dstRepo restic.Repository, treeID restic.ID) error {
+	tree, err := srcRepo.LoadTree(ctx, treeID)
+	if err != nil {
+		return fmt.Errorf("LoadTree(%v) returned error %v", treeID.Str(), err)
+	}
+	// Do we already have this tree blob?
+	if !dstRepo.Index().Has(treeID, restic.TreeBlob) {
```

Comment:
IMO, you can optimize here: if `dstRepo` does have this `TreeBlob`, it should also have all entries (and sub-trees). 
At least if this repository is in a sane state :wink:  So no need to run into the recursion!

Maybe we could also add a flag `--deep-copy` to always recurse into the trees. This would optionally allow to "heal" missing blobs if they are present in another repo/snapshot which is selected for copy - at the cost of copy speed.

However, here we do have the same problem with `Index.Has` as @MichaelEischer stated below.

Reply:
I don't think that this definition of "sane" is a good idea. Using it, just interrupting a backup would turn a repository "insane" which I regard as highly problematic (especially for mobile clients). It would also be rather complex to implement prune in a way that can for every repository guarantee, that it never enters an "insane" state. The trees in rewritten packs would have to be laid out such that dependencies between tree packs are acyclic (a tree pack `p` depends on all packs which contain tree blobs referenced by tree blobs in `p`). If we want to avoid rewriting every tree pack, then the complexity just skyrockets. It's actually a design goal of prune to never leave the repository in a broken state (requiring repair steps beyond removing the stale lock) no matter when it is interrupted.

The definition of "sane" used by the `check` command is rather simple: A repository is "sane", if every snapshot can restored.

As restic tries to prevent users from accidentally damaging their backup, requiring a "sane" and pruned repository would require the copy command to run a _full_ check of the destination repository which sabotages the goal of speeding up the copy command.

Reply:
Actually after the `vistitedTrees` optimization there is not that much to gain here by skipping the reading of trees already present in the destination repo. Yes, it will walk the tree once but detect that there is nothing to save and hence it will be as faster than a `backup` run with no modified files. Coping multiple snapshots with similar tree structure will get sped up even more by using `visitedTrees`. That should be fine :+1: 

You are right that it wouldn't help to define a "sane" state which can be only obtained safely by always pruning. Nevertheless I think a good definition of a "sane" state (and the pitfalls like being able to have "dead" trees as long as they themselves are not referenced by snapshots), maybe also the definition of an "optimal" state (in terms of e.g. not wasting space), would be good to have. This could also allow `check` to produce a summary like "your repository is sane but not optimal". But I'll stop now spoiling this PR with this unrelated topic :wink: